crates/api_common/src/request.rs

```diff
@@ -65,95 +66,97 @@ pub async fn fetch_link_metadata(url: &Url, context: &LemmyContext) -> LemmyResu
+    if allow_generate_thumbnail {
+      match post.url {
+        Some(url) => generate_pictrs_thumbnail(&url, &context)
+          .await
+          .ok()
+          .map(Into::into),
+        None => None,
+      }
+    } else {
+      None
     }
-    // Generate local thumbnail if allowed
-    else if allow_generate_thumbnail {
-      match post
-        .url
-        .filter(|_| is_image_post)
-        .or(metadata.opengraph_data.image)
-      {
-        Some(url) => generate_pictrs_thumbnail(&url, &context).await.ok(),
+  } else {
+    // Use custom thumbnail if available and its not an image post
+    if let Some(custom_thumbnail) = custom_thumbnail {
+      proxy_image_link(custom_thumbnail, &context).await.ok()
+    } else if allow_generate_thumbnail {
+      match metadata.opengraph_data.image {
+        Some(url) => generate_pictrs_thumbnail(&url, &context)
+          .await
+          .ok()
+          .map(Into::into),
         None => None,
       }
     }
     // Otherwise use opengraph preview image directly
     else {
-      metadata.opengraph_data.image.map(Into::into)
-    };
+      metadata.opengraph_data.image
+    }
+  };
 
-    // Proxy the image fetch if necessary
-    let proxied_thumbnail_url = proxy_image_link_opt_apub(thumbnail_url, &context).await?;
+  let form = PostUpdateForm {
+    embed_title: Some(metadata.opengraph_data.title),
+    embed_description: Some(metadata.opengraph_data.description),
+    embed_video_url: Some(metadata.opengraph_data.embed_video_url),
+    thumbnail_url: Some(thumbnail_url),
+    url_content_type: Some(metadata.content_type),
+    ..Default::default()
+  };
+  let updated_post = Post::update(&mut context.pool(), post.id, &form).await?;
+  if let Some(send_activity) = send_activity(updated_post) {
+    ActivityChannel::submit_activity(send_activity, &context).await?;
+  }
+  Ok(())
+}
 
-    let form = PostUpdateForm {
-      embed_title: Some(metadata.opengraph_data.title),
-      embed_description: Some(metadata.opengraph_data.description),
-      embed_video_url: Some(metadata.opengraph_data.embed_video_url),
-      thumbnail_url: Some(proxied_thumbnail_url),
-      url_content_type: Some(metadata.content_type),
-      ..Default::default()
-    };
-    let updated_post = Post::update(&mut context.pool(), post.id, &form).await?;
-    if let Some(send_activity) = send_activity(updated_post) {
-      ActivityChannel::submit_activity(send_activity, &context).await?;
-    }
-    Ok(())
-  });
+/// Generates a post thumbnail in background task, because some sites can be very slow to respond.
+pub fn generate_post_link_metadata_background(
+  post: Post,
+  custom_thumbnail: Option<Url>,
+  send_activity: impl FnOnce(Post) -> Option<SendActivityData> + Send + 'static,
+  local_site: Option<LocalSite>,
+  context: Data<LemmyContext>,
+) {
+  spawn_try_task(async move {
+    generate_post_link_metadata(post, custom_thumbnail, send_activity, local_site, context).await
+  })
```

Comment:
This is only called in a single place, so remove the wrapper function and instead put `spawn_try_task()` directly.

Reply:
Sure, I'll do that now.